### PR TITLE
Node factory

### PR DESCRIPTION
**Pull request by J-Loudet**

### CODE REVIEW

File: zenoh-flow/src/model/record/dataflow.rs [hunk 1]
```diff
@@ -36,9 +36,9 @@ pub struct DataFlowRecord {
     pub operators: HashMap<NodeId, OperatorRecord>,
     pub sinks: HashMap<NodeId, SinkRecord>,
     pub sources: HashMap<NodeId, SourceRecord>,
-    pub connectors: Vec<ZFConnectorRecord>,
+    pub connectors: HashMap<NodeId, ZFConnectorRecord>,
     pub links: Vec<LinkRecord>,
-    pub counter: u32,
+    pub counter: usize,
```
Comment (gabrik) [inline zenoh-flow/src/model/record/dataflow.rs]: It was `u32` to somehow limit the number of nodes and links in a graph.

If we are happy to have `usize` as this limit I'm fine with it.

File: zenoh-flow/src/model/record/dataflow.rs [hunk 2]
```diff
@@ -248,19 +248,18 @@ impl DataFlowRecord {
...
                     let sender_id = format!(
                         "sender-{}-{}-{}-{}",
                         &self.flow, &self.uuid, &l.from.node, &l.from.output
-                    );
+                    )
+                    .into();
                     let sender = ZFConnectorRecord {
                         kind: ZFConnectorKind::Sender,
-                        id: sender_id.clone().into(),
+                        id: Arc::clone(&sender_id),
```
Comment (gabrik) [inline zenoh-flow/src/model/record/dataflow.rs]: Unnecessary, if `sender_id` is already `Arc<str>` so just `sender_id.clone()` works.
If used like this we "leak" the fact that the ID is behind an `Arc`

File: zenoh-flow/src/model/record/dataflow.rs [hunk 3]
```diff
@@ -276,7 +275,7 @@ impl DataFlowRecord {
                     let link_sender = LinkDescriptor {
                         from: l.from.clone(),
                         to: InputDescriptor {
-                            node: sender_id.into(),
+                            node: Arc::clone(&sender_id),
```
Comment (gabrik) [inline zenoh-flow/src/model/record/dataflow.rs]: Like before.

File: zenoh-flow/src/model/record/dataflow.rs [hunk 4]
```diff
@@ -294,10 +293,11 @@ impl DataFlowRecord {
                 let receiver_id = format!(
                     "receiver-{}-{}-{}-{}",
                     &self.flow, &self.uuid, &l.to.node, &l.to.input
-                );
+                )
+                .into();
                 let receiver = ZFConnectorRecord {
                     kind: ZFConnectorKind::Receiver,
-                    id: receiver_id.clone().into(),
+                    id: Arc::clone(&receiver_id),
```
Comment (gabrik) [inline zenoh-flow/src/model/record/dataflow.rs]: Same as before.

File: zenoh-flow/src/model/record/dataflow.rs [hunk 5]
```diff
@@ -312,7 +312,7 @@ impl DataFlowRecord {
                 // Creating link between receiver and node
                 let link_receiver = LinkDescriptor {
                     from: OutputDescriptor {
-                        node: receiver_id.into(),
+                        node: Arc::clone(&receiver_id),
```
Comment (gabrik) [inline zenoh-flow/src/model/record/dataflow.rs]: Same as before.

File: zenoh-flow/src/runtime/dataflow/instance/mod.rs [hunk 6]
```diff
@@ -14,39 +14,89 @@
...
 use crate::Result;
 use std::collections::HashMap;
+use std::ops::Deref;
 use std::sync::Arc;
 use uhlc::HLC;
-use uuid::Uuid;
 
-use self::runners::Runner;
+/// TODO(J-Loudet) Improve documentation.
+pub struct DataFlowInstance {
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/instance/mod.rs:37]: Maybe keep this for another PR. :)

File: zenoh-flow/src/runtime/dataflow/instance/runners/mod.rs [hunk 7]
```diff
@@ -39,27 +45,207 @@ pub enum RunAction {
...
 
-/// This traits abstracts the runners, it provides the functions that are
-/// common across the runners.
+/// TODO(J-Loudet) Improve documentation.
 ///
-///
-#[async_trait]
-pub trait Runner: Send + Sync {
-    /// The actual run where the magic happens.
+/// A Runner2 takes care of running a node. Regardless of what it is. No need for a trait, no need
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/instance/runners/mod.rs]: `s/Runner2/Runner/g`

File: zenoh-flow/src/runtime/dataflow/instance/runners/mod.rs [hunk 8]
```diff
@@ -39,27 +45,207 @@ pub enum RunAction {
...
+            let (abort_handle, abort_registration) = AbortHandle::new_pair();
+            let handle = async_std::task::spawn(Abortable::new(run_loop, abort_registration));
+
+            self.run_loop_handle = Some(handle);
+            self.run_loop_abort_handle = Some(abort_handle);
+        }
+
+        if !self.outputs_callbacks.is_empty() {
+            // TODO How is that clone "safe"? The underlying channels are "job stealing" so only the
+            // last copy actually gets the data?
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/instance/runners/mod.rs]: It is not the last that receive the data, but the one that is ready on the `recv_async().await`

File: zenoh-flow/src/runtime/dataflow/instance/runners/mod.rs [hunk 9]
```diff
@@ -39,27 +45,207 @@ pub enum RunAction {
...
+    ///       won’t have access to the CPU for these 10 seconds.
+    ///       A possible fix would be to spawn a thread per callback. What’s the impact?
+    pub(crate) fn start(&mut self) {
+        if self.is_running() {
+            log::warn!("Called `start` while node is ALREADY running. Returning.");
+            return;
+        }
+
+        if let Some(node) = &self.node {
+            let node = Arc::clone(node);
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/instance/runners/mod.rs]: `let node = node.clone()`

File: zenoh-flow/src/runtime/dataflow/mod.rs [hunk 10]
```diff
@@ -18,275 +18,331 @@ pub mod node;
...
-        &mut self,
-        id: NodeId,
-        configuration: Option<Configuration>,
-        _outputs: Vec<PortDescriptor>,
-        source: Arc<dyn Source>,
-    ) -> ZFResult<()> {
-        // self.validator.try_add_source(id.clone(), output.clone())?;
-        // self.validator.try_add_source(id.clone(), &outputs)?;
+    /// Try instantiating the DataFlow2, hence generating a DataFlowInstance2.
+    pub async fn try_instantiate(self, hlc: Arc<HLC>) -> ZFResult<DataFlowInstance> {
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/mod.rs]: Can you leave these changes for another PR? :)

**Review by gabrik — CHANGES_REQUESTED** [review]
I think some code should be part of another PR.
Like the refactoring of `DataFlow` and `DataFlowInstance`

File: zenoh-flow/src/model/record/dataflow.rs (hunk 1 above)
Comment (J-Loudet) [reply zenoh-flow/src/model/record/dataflow.rs]: I believe `usize` is preferable when working with integers (as it will have the size of a machine word for the target architecture) which is why I went through with that modification. However, at least in this context, the performance gains are relatively limited (if there are any).

The limit should be implemented through some other means. I’ll revert the changes as the modification could be done the moment we decide to enforce a limit on the number of nodes / links.

File: zenoh-flow/src/model/record/dataflow.rs (hunk 2 above)
Comment (J-Loudet) [reply zenoh-flow/src/model/record/dataflow.rs]: I prefer to have an explicit call to `Arc::clone` as we know that it is just a reference increment and not an expensive operation.
I’ll revert this changes as, as you pointed out, it can be hidden in the "newtype" we intend on creating.

File: zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs [hunk 11]
```diff
@@ -13,326 +13,124 @@
...
-    pub(crate) handle: Option<JoinHandle<Result<ZFResult<()>, Aborted>>>,
-    pub(crate) abort_handle: Option<AbortHandle>,
-    pub(crate) link: Input,
+    pub(crate) input: Input,
+    pub(crate) z_session: Arc<zenoh::Session>,
     pub(crate) key_expr: KeyExpr<'static>,
 }
 
 impl ZenohSender {
-    /// Creates a new `ZenohSender` with the given parameters.
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs:41]: Put back the docstrings pls :)

File: zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs [hunk 12]
```diff
@@ -13,326 +13,124 @@
...
+impl Node for ZenohSender {
+    async fn iteration(&self) -> ZFResult<()> {
+        if let Ok(message) = self.input.recv_async().await {
+            log::trace!("[ZenohSender: {}] recv_async: OK", self.id);
+
+            let serialized = message.serialize_bincode()?;
+
+            self.z_session
+                .put(self.key_expr.clone(), serialized)
+                // .put(key_expr, &(**buffer)[0..size])
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs]: Remove this line

File: zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs [hunk 13]
```diff
@@ -13,326 +13,124 @@
...
+                .await
+        } else {
+            Err(zferror!(ErrorKind::Disconnected).into())
         }
-
-        Ok(())
     }
 }
 
-/// A `ZenohReceiver` receives the messages from Zenoh when nodes are running
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs:188]: Keep the docs please

File: zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs [hunk 14]
```diff
@@ -13,326 +13,124 @@
...
 use async_trait::async_trait;
-use futures::stream::{AbortHandle, Abortable, Aborted};
+use flume::Receiver;
 use std::sync::Arc;
-use zenoh::prelude::*;
+use zenoh::prelude::r#async::*;
+use zenoh::subscriber::Subscriber;
+use zenoh_util::core::AsyncResolve;
 
-/// The `ZenohSender` is the connector that sends the data to Zenoh
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs:28]: Keep docs

File: zenoh-flow/src/runtime/dataflow/loader.rs [hunk 15]
```diff
@@ -343,85 +337,56 @@ impl Loader {
...
-            .get::<*mut OperatorDeclaration>(b"zfoperator_declaration\0")?
-            .read();
-
-        // version checks to prevent accidental ABI incompatibilities
-        if decl.rustc_version != RUSTC_VERSION || decl.core_version != CORE_VERSION {
-            return Err(zferror!(ErrorKind::VersionMismatch).into());
-        }
-
-        Ok((library, (decl.register)()?))
+    unsafe fn load_lib_source_factory(
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/loader.rs:364]: Keep docs on this function

**Review by gabrik — CHANGES_REQUESTED** [review]
Keep the doc when it is already there.

**Review by gabrik — APPROVED** [review]
